tech-spec.md

```diff
@@ -0,0 +1,310 @@
+
+### Private modules
+
+#### cons
+
+Data source connection abstraction, including intermediate copy of spreadsheet
+files to in-memory DBs.
+
+- connect(data_source: str) -> Connection
+
+#### rules
+
+Loading of rule files.
+
+- load(schema_file: str) -> Dict[RuleId, Rule]
+
+#### trees
+
+Parsing of rules into abstract syntax trees.
+
+- parse(rules: Dict[RuleId, Rule] | List[Rule]) -> RuleTree
+
+#### queries
+
+(SQL) query generation from ASTs.
+
+- generate(rt: RuleTree) -> Dict[OrgName, Query]
+
+### Examples
+
+common definitions:
+```python
+import pandas as pd
+
+import sharing as s
+
+data_file = 'data.xlsx'
+rule_file = 'rules.csv'
+org = 'OHRI'
+```
+
+high-level one-shot function:
+
+```python
+results = s.extract(data_file, rules, org)
+for table, data in results.items():
+    data.to_csv(f'{org}-{table}.csv')
+```
+
+low-level sample code:
+
+```python
+def describeTableQuery(con, rules, table, query):
+    print(f'query table: {table}')
+
+    (selectRuleId, columns) = s.getColumns(con, query)
+    print(f'query columns (from rule {selectRuleId}):')
+    print(','.join(columns))
+
+    print('query counts per rule:')
+    ruleCounts = s.getCounts(con, query)
+    for ruleId, count in ruleCounts.items():
+        r = rules[ruleId]
+        ruleFilter = f'{r.key} {r.operator} {r.value}'
+        print(f'{ruleId} | {count} | {ruleFilter}')
+
+def extractFilteredData(con, table, query):
+    data: pd.Dataframe = s.extract(con, query)
+    data.to_csv(f'{org}-{table}.csv')
+
+con = s.connect(data_file)
+rules = s.load(rule_file)
+ruleTree = s.parse(rules)
+tableQueries = s.generate(ruleTree)[org]
+for table, query in tableQueries.items():
+    describeTableQuery(con, rules, table, query)
+    extractFilteredData(con, table, query)
+```
+
+## Rule schema parsing
+
+1. open schema file
+2. read lines
+3. parse each line into a rule obj
+4. add each rule obj to a dictionary with rule id as key
+
+## AST generation
+
+Node kinds:
+- **root**: AST root
+- **share**: what to share with each org
+- **table**: table name
+- **select**: lists column name values
+- **group**: groups filters together
+- **filter**: defines a filter with operator, key and value
+- **key**: field name
+- **value**: generic
+
+Node structure:
+
+- (ruleId: int)
+- kind: NodeKind
+- value: str
+- children: List[Node]
+
+For each rule, a node is added based on its mode:
+- select:
+    - (**select**, ("" or "all")):
+        - (**value**, column) for column in rule.key
+- filter:
+    - (**filter**, rule.operator):
+        - (**key**, rule.key)
+        - (**value**, x) for x in rule.value
+- group:
+    - (**group**, rule.operator):
+        - nodes where node.ruleId in rule.value
+- share:
+    - (**root**, ""):
+        - (**share**, org) for org in rule.key
+            - (**table**, x) for x in select-rule-tables
+                - filter-root node
+
+Example rules with its generated tree:
+
+```csv
+ruleId,table,mode,key,operator,value,notes
```

Comment:
Can you format this as a markdown table? So its easier to read in the rendered HTML.